programs/host/frontend.c

```diff
@@ -0,0 +1,943 @@
+            switch( mode )
+            {
+            case MBEDTLS_SERIALIZE_BLOCK_BLOCK:
+                ret = mbedtls_net_set_block( &ctx );
+                break;
+            case MBEDTLS_SERIALIZE_BLOCK_NONBLOCK:
+                ret = mbedtls_net_set_nonblock( &ctx );
+                break;
+            default:
+                ret = MBEDTLS_ERR_SERIALIZE_BAD_INPUT;
+                break;
+            }
+        }
+        break;
+
+    case MBEDTLS_SERIALIZE_FUNCTION_RECV:
+        CHECK_ARITY( 3 );
+        CHECK_LENGTH( 0, 2 ); // fd
+        CHECK_LENGTH( 1, 4 ); // len
+        CHECK_LENGTH( 2, 4 ); // timeout
+        {
+            mbedtls_net_context ctx = { item_int16( inputs[0] ) };
+            uint32_t len = item_int32( inputs[1] );
+            uint32_t timeout = item_int32( inputs[2] );
+            ALLOC_OUTPUT( 0 , len ); // data
+            if( timeout == MBEDTLS_SERIALIZE_TIMEOUT_INFINITE )
+            {
+                DBG( "executing recv fd=%u len=%u", (int) ctx.fd, (unsigned) len );
+                ret = mbedtls_net_recv( &ctx, item_buffer( outputs[0] ), len );
+            }
+            else
+            {
+                DBG( "executing recv_timeout fd=%u len=%u timeout=%u", (unsigned) ctx.fd, (unsigned) len, (unsigned) timeout );
+                ret = mbedtls_net_recv_timeout( &ctx,
+                                                item_buffer( outputs[0] ), len,
+                                                timeout );
+            }
+            if( ret >= 0 )
+            {
+                DBG( "received %zu bytes on fd=%d", (size_t) ret, (int) ctx.fd );
+                outputs[0]->size = ret;
+                ret = 0;
+            }
+        }
+        break;
+
+    case MBEDTLS_SERIALIZE_FUNCTION_SEND:
+        CHECK_ARITY( 2 );
+        CHECK_LENGTH( 0, 2 ); // fd
+        {
+            mbedtls_net_context ctx = { item_int16( inputs[0] ) };
+            size_t len = inputs[1]->size;
+            unsigned char *buf = item_buffer( inputs[1] );
+            ALLOC_OUTPUT( 0 , 4 ); // sent_len
+            DBG( "executing send fd=%u len=%zu", (int) ctx.fd, len );
+            ret = mbedtls_net_send( &ctx, buf, len );
+            if( ret >= 0 )
+            {
+                DBG( "sent %zu bytes on fd=%d", (size_t) ret, (int) ctx.fd );
+                set_item_int32( outputs[0], ret );
+                ret = 0;
+            }
+        }
+        break;
+
+    case MBEDTLS_SERIALIZE_FUNCTION_SHUTDOWN:
+        CHECK_ARITY( 1 );
+        CHECK_LENGTH( 0, 2 ); // fd
+        {
+            mbedtls_net_context ctx = { item_int16( inputs[0] ) };
+            DBG( "executing shutdown fd=%d", (int) ctx.fd );
+            mbedtls_net_free( &ctx );
+            ret = 0;
+        }
+        break;
+
+    case MBEDTLS_SERIALIZE_FUNCTION_FOPEN:
+        {
+            char * path = NULL;
+            char * mode = NULL;
+            int32_t file_id;
+
+            ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
```

Comment:
Don't use `MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT` in the frontend. This error code indicates that something wrong happened when the target tried to deserialize the outputs. The frontend should return an appropriate status for the error (define some new `MBEDTLS_ERR_IO` constants that correspond to common `errno` vaues?), or `MBEDTLS_ERR_SERIALIZE_UNSUPPORTED_OUTPUT` if the operation succeeded but then something went wrong while trying to serialize the result.